File: recipes/hostui/gudule-greeting-hostui-qt/src/DaemonProcess.cpp
#include "DaemonProcess.h"

#include <QCoreApplication>
#include <QDir>
#include <QElapsedTimer>
#include <QFile>
#include <QFileInfo>
#include <QRegularExpression>
#include <QTextStream>

#include <set>
#include <stdexcept>

#include "holons/holons.hpp"

namespace {
constexpr auto kBinaryPrefix = "gudule-daemon-greeting-";
constexpr auto kHolonUUID = "77faf86e-e778-4cbc-a484-c545a0e35ce3";

QString displayVariant(const QString &variant) {
  const QStringList tokens = variant.split('-', Qt::SkipEmptyParts);
  QStringList parts;
  parts.reserve(tokens.size());
  for (const QString &token : tokens) {
    if (token == QStringLiteral("cpp")) {
      parts << QStringLiteral("CPP");
    } else if (token == QStringLiteral("js")) {
      parts << QStringLiteral("JS");
    } else if (token == QStringLiteral("qt")) {
      parts << QStringLiteral("Qt");
    } else if (!token.isEmpty()) {
      QString title = token;
      title[0] = title.at(0).toUpper();
      parts << title;
    }
  }
  return parts.join('-');
}

QString buildRunnerFor(const QString &variant) {
  if (variant == QStringLiteral("go")) {
    return QStringLiteral("go-module");
  }
  if (variant == QStringLiteral("rust")) {
    return QStringLiteral("cargo");
  }
  if (variant == QStringLiteral("swift")) {
    return QStringLiteral("swift-package");
  }
  if (variant == QStringLiteral("kotlin")) {
    return QStringLiteral("gradle");
  }
  if (variant == QStringLiteral("dart")) {
    return QStringLiteral("dart");
  }
  if (variant == QStringLiteral("python")) {
    return QStringLiteral("python");
  }
  if (variant == QStringLiteral("csharp")) {
    return QStringLiteral("dotnet");
  }
  if (variant == QStringLiteral("node")) {
    return QStringLiteral("npm");
  }
  return QStringLiteral("go-module");
}

std::optional<DaemonProcess::GreetingDaemonIdentity> identityFromBinaryPath(
    const QString &binaryPath) {
  const QString fileName = QFileInfo(binaryPath).fileName();
  QString normalized = fileName;
  if (normalized.endsWith(QStringLiteral(".exe"))) {
    normalized.chop(4);
  }
  if (!normalized.startsWith(QString::fromUtf8(kBinaryPrefix))) {
    return std::nullopt;
  }

  const QString variant = normalized.mid(qstrlen(kBinaryPrefix));
  return DaemonProcess::GreetingDaemonIdentity{
      QStringLiteral("gudule-greeting-daemon-%1").arg(variant),
      QStringLiteral("Greeting-Daemon-%1").arg(displayVariant(variant)),
      normalized,
      buildRunnerFor(variant),
      QFileInfo(binaryPath).absoluteFilePath(),
  };
}

void addBundledBinaries(const QString &directoryPath, QStringList *results,
                        std::set<QString> *seen) {
  const QDir directory(directoryPath);
  if (!directory.exists()) {
    return;
  }

  const QFileInfoList entries =
      directory.entryInfoList(QStringList(QStringLiteral("%1*").arg(QString::fromUtf8(kBinaryPrefix))),
                              QDir::Files | QDir::NoDotAndDotDot, QDir::Name);
  for (const QFileInfo &entry : entries) {
    const QString normalized = entry.absoluteFilePath();
    if (seen->insert(normalized).second) {
      results->append(normalized);
    }
  }
}

void addSourceTreeDaemons(const QString &directoryPath, QStringList *results,
                          std::set<QString> *seen) {
  const QDir daemonsDir(directoryPath);
  if (!daemonsDir.exists()) {
    return;
  }

  const QFileInfoList entries = daemonsDir.entryInfoList(
      QStringList(QStringLiteral("%1*").arg(QString::fromUtf8(kBinaryPrefix))),
      QDir::Dirs | QDir::NoDotAndDotDot, QDir::Name);
  for (const QFileInfo &entry : entries) {
    const QString binaryName = entry.fileName();
    const QString built =
        QDir(entry.absoluteFilePath()).filePath(QStringLiteral(".op/build/bin/%1").arg(binaryName));
    const QString local = QDir(entry.absoluteFilePath()).filePath(binaryName);
    if (seen->insert(built).second) {
      results->append(QDir::cleanPath(built));
    }
    if (seen->insert(local).second) {
      results->append(QDir::cleanPath(local));
    }
  }
}

QString firstUri(const QString &line) {
  static const QRegularExpression re(QStringLiteral(R"(tcp://\S+)"));
  const QRegularExpressionMatch match = re.match(line);
  if (!match.hasMatch()) {
    return {};
  }
  return match.captured(0);
}

QString assemblyFamily() {
  return qEnvironmentVariable("OP_ASSEMBLY_FAMILY", "Greeting-Qt-Go");
}

QString assemblyTransport() {
  return qEnvironmentVariable("OP_ASSEMBLY_TRANSPORT", "tcp");
}

QString displayConnectionTarget(const QString &value) {
  QString trimmed = value.trimmed();
  for (const auto &prefix :
       {QStringLiteral("tcp://"), QStringLiteral("http://"),
        QStringLiteral("https://"), QStringLiteral("ws://"),
        QStringLiteral("wss://"), QStringLiteral("stdio://")}) {
    if (trimmed.startsWith(prefix, Qt::CaseInsensitive)) {
      return trimmed.mid(prefix.size());
    }
  }
  return trimmed;
}

void logHostUI(const QString &line) {
  QTextStream stream(stderr);
  stream << line << Qt::endl;
}
}  // namespace

DaemonProcess::DaemonProcess(QObject *parent) : QObject(parent) {
}

bool DaemonProcess::start() {
  if (channel_) {
    return true;
  }
  lastError_.clear();
  grpcTarget_.clear();

  const auto daemon = resolveDaemon();
  if (!daemon.has_value()) {
    lastError_ = QStringLiteral("Daemon binary not found: %1*")
                     .arg(QString::fromUtf8(kBinaryPrefix));
    return false;
  }
  binaryPath_ = daemon->binaryPath;
  daemonSlug_ = daemon->slug;
  logHostUI(QStringLiteral("[HostUI] assembly=%1 daemon=%2 transport=%3")
                .arg(assemblyFamily(), daemon->binaryName, assemblyTransport()));

  auto stageRoot = std::make_unique<QTemporaryDir>(
      QDir::temp().filePath(QStringLiteral("gudule-greeting-hostui-qt-XXXXXX")));
  if (!stageRoot->isValid()) {
    lastError_ = QStringLiteral("Failed to create temporary holon root.");
    return false;
  }

  const QDir root(stageRoot->path());
  const QString holonDir =
      root.filePath(QStringLiteral("holons/%1").arg(daemon->slug));
  if (!QDir().mkpath(holonDir)) {
    lastError_ = QStringLiteral("Failed to create staged holon directory: %1").arg(holonDir);
    return false;
  }

  QFile manifest(QDir(holonDir).filePath(QStringLiteral("holon.yaml")));
  if (!manifest.open(QIODevice::WriteOnly | QIODevice::Text | QIODevice::Truncate)) {
    lastError_ = QStringLiteral("Failed to write staged holon manifest: %1")
                     .arg(manifest.errorString());
    return false;
  }
  manifest.write(buildManifest(*daemon).toUtf8());
  manifest.close();

  QString portFilePath;
  try {
    portFilePath = startBundledDaemon(*daemon, stageRoot->path());
  } catch (const std::exception &ex) {
    lastError_ = QString::fromUtf8(ex.what());
    if (daemonProcess_) {
      daemonProcess_->kill();
      daemonProcess_->waitForFinished(1000);
      daemonProcess_.reset();
    }
    return false;
  }

  const QString previousDirectory = QDir::currentPath();
  if (!QDir::setCurrent(stageRoot->path())) {
    lastError_ = QStringLiteral("Failed to enter staged holon root: %1").arg(stageRoot->path());
    daemonProcess_->kill();
    daemonProcess_->waitForFinished(1000);
    daemonProcess_.reset();
    return false;
  }

  try {
    holons::ConnectOptions opts;
    opts.transport = assemblyTransport().toStdString();
    opts.start = false;
    opts.port_file = portFilePath.toStdString();
    channel_ = holons::connect(daemon->slug.toStdString(), opts);
    grpcTarget_ = QString::fromStdString(holons::channel_target(channel_));
    if (grpcTarget_.isEmpty()) {
      throw std::runtime_error("cpp-holons did not expose the daemon target");
    }
    logHostUI(QStringLiteral("[HostUI] connected to %1 on %2")
                  .arg(daemon->binaryName, displayConnectionTarget(grpcTarget_)));
    stageRoot_ = std::move(stageRoot);
  } catch (const std::exception &ex) {
    lastError_ = QString::fromUtf8(ex.what());
    if (channel_) {
      try {
        holons::disconnect(channel_);
      } catch (const std::exception &) {
      }
      channel_.reset();
    }
    grpcTarget_.clear();
    if (daemonProcess_) {
      daemonProcess_->kill();
      daemonProcess_->waitForFinished(1000);
      daemonProcess_.reset();
    }
  }

  QDir::setCurrent(previousDirectory);
  return static_cast<bool>(channel_);
}

void DaemonProcess::stop() {
  if (channel_) {
    try {
      holons::disconnect(channel_);
    } catch (const std::exception &ex) {
      lastError_ = QString::fromUtf8(ex.what());
    }
    channel_.reset();
  }

  if (daemonProcess_) {
    daemonProcess_->kill();
    daemonProcess_->waitForFinished(1000);
    daemonProcess_.reset();
  }

  stageRoot_.reset();
  daemonSlug_.clear();
  binaryPath_.clear();
  grpcTarget_.clear();
}

std::shared_ptr<grpc::Channel> DaemonProcess::channel() const { return channel_; }

QString DaemonProcess::target() const { return daemonSlug_; }

QString DaemonProcess::grpcTarget() const { return grpcTarget_; }

QString DaemonProcess::binaryPath() const { return binaryPath_; }

QString DaemonProcess::lastError() const { return lastError_; }

std::optional<DaemonProcess::GreetingDaemonIdentity> DaemonProcess::resolveDaemon() const {
  QStringList candidates;
  std::set<QString> seen;
  const QDir currentDir(QDir::currentPath());
  addBundledBinaries(currentDir.filePath(QStringLiteral("build")), &candidates, &seen);
  addBundledBinaries(currentDir.filePath(QStringLiteral("../build")), &candidates, &seen);
  addSourceTreeDaemons(currentDir.filePath(QStringLiteral("../../daemons")), &candidates, &seen);

  const QDir appDir(QCoreApplication::applicationDirPath());
  addBundledBinaries(appDir.path(), &candidates, &seen);
  addBundledBinaries(appDir.filePath(QStringLiteral("daemon")), &candidates, &seen);
  addBundledBinaries(appDir.filePath(QStringLiteral("../Resources")), &candidates, &seen);
  addBundledBinaries(appDir.filePath(QStringLiteral("../Resources/daemon")), &candidates, &seen);

  for (const QString &candidate : candidates) {
    const QFileInfo info(candidate);
    if (!info.exists() || !info.isFile() || !info.isExecutable()) {
      continue;
    }
    if (const auto identity = identityFromBinaryPath(candidate)) {
      return identity;
    }
  }

  return std::nullopt;
}

QString DaemonProcess::startBundledDaemon(const GreetingDaemonIdentity &daemon,
                                          const QString &stageRootPath) {
  auto process = std::make_unique<QProcess>();
  process->setProgram(daemon.binaryPath);
  process->setArguments(
      {QStringLiteral("serve"), QStringLiteral("--listen"), QStringLiteral("tcp://127.0.0.1:0")});
  process->setProcessChannelMode(QProcess::MergedChannels);
  process->start();
  if (!process->waitForStarted()) {
    throw std::runtime_error("Failed to launch bundled daemon");
  }

  QString listenUri;
  QStringList recentLines;
  QElapsedTimer timer;
  timer.start();
  QByteArray buffered;

  while (timer.elapsed() < 5000 && listenUri.isEmpty()) {
    process->waitForReadyRead(100);
    buffered.append(process->readAllStandardOutput());
    const QList<QByteArray> lines = buffered.split('\n');
    buffered = lines.isEmpty() ? QByteArray() : lines.last();
    for (int i = 0; i + 1 < lines.size(); ++i) {
      const QString line = QString::fromUtf8(lines.at(i)).trimmed();
      if (line.isEmpty()) {
        continue;
      }
      recentLines << line;
      if (recentLines.size() > 8) {
        recentLines.removeFirst();
      }
      const QString candidate = firstUri(line);
      if (!candidate.isEmpty()) {
        listenUri = candidate;
        break;
      }
    }

    if (process->state() == QProcess::NotRunning && listenUri.isEmpty()) {
      throw std::runtime_error(
          QStringLiteral("Bundled daemon exited before advertising an address: %1")
              .arg(recentLines.join(QStringLiteral(" | ")))
              .toStdString());
    }
  }

  if (listenUri.isEmpty()) {
    process->kill();
    process->waitForFinished(1000);
    throw std::runtime_error("Bundled daemon did not advertise a tcp:// address");
  }

  const QString portFilePath =
      QDir(stageRootPath).filePath(QStringLiteral(".op/run/%1.port").arg(daemon.slug));
  QDir().mkpath(QFileInfo(portFilePath).absolutePath());
  QFile portFile(portFilePath);
  if (!portFile.open(QIODevice::WriteOnly | QIODevice::Text | QIODevice::Truncate)) {
    process->kill();
    process->waitForFinished(1000);
    throw std::runtime_error("Failed to write daemon port file");
  }
  QTextStream(&portFile) << listenUri << '\n';
  portFile.close();

  daemonProcess_ = std::move(process);
  return portFilePath;
}

QString DaemonProcess::buildManifest(const GreetingDaemonIdentity &daemon) const {
  QString escapedPath = daemon.binaryPath;
  escapedPath.replace(QStringLiteral("\\"), QStringLiteral("\\\\"));
  escapedPath.replace(QStringLiteral("\""), QStringLiteral("\\\""));

  return QStringLiteral(
             "schema: holon/v0\n"
             "uuid: \"%1\"\n"
             "given_name: gudule\n"
             "family_name: \"%2\"\n"
             "motto: Greets users in 56 languages through the bundled daemon.\n"
             "composer: Codex\n"
             "clade: deterministic/pure\n"
             "status: draft\n"
             "born: \"2026-03-12\"\n"
             "generated_by: manual\n"
             "kind: native\n"
             "build:\n"
             "  runner: %3\n"
             "artifacts:\n"
             "  binary: \"%4\"\n")
      .arg(QString::fromUtf8(kHolonUUID), daemon.familyName, daemon.buildRunner, escapedPath);
}

File: recipes/hostui/gudule-greeting-hostui-swiftui/GreetingSwiftUI/DaemonProcess.swift
import Foundation
#if os(macOS)
import Holons
#endif

/// Manages the lifecycle of the bundled greeting daemon connection.
@MainActor
final class DaemonProcess: ObservableObject {
    @Published var isRunning = false
    @Published var connectionError: String?

    private var client: GreetingClient?
#if os(macOS)
    private var stageRoot: URL?
    private var daemonIdentity: GreetingDaemonIdentity?
#endif

    func start() {
        guard client == nil else { return }
        connectionError = nil

        if transport == "mem" {
            let family = assemblyFamily.lowercased()
            let parts = family.split(separator: "-")
            if parts.count >= 3 {
                let ui = parts[1]
                let daemon = String(parts[2].prefix(while: { $0.isLetter }))
                let uiLang = ui.hasPrefix("swift") ? "swift" : String(ui)
                if uiLang != daemon {
                    connectionError = "memory connection mode requires the same language for UI and the daemon"
                    isRunning = false
                    return
                }
            }
        }

#if os(macOS)
        do {
            let daemon = try resolveDaemon()
            daemonIdentity = daemon
            let root = try stageHolonRoot(daemon: daemon)
            stageRoot = root

            let previousDirectory = FileManager.default.currentDirectoryPath
            guard FileManager.default.changeCurrentDirectoryPath(root.path) else {
                throw DaemonStartError.failedToEnterRoot(root.path)
            }
            defer {
                FileManager.default.changeCurrentDirectoryPath(previousDirectory)
            }

            logHostUI("[HostUI] assembly=\(assemblyFamily) daemon=\(daemon.binaryName) transport=\(transport)")
            var options = ConnectOptions()
            options.transport = transport
            
            client = try GreetingClient.connected(to: daemon.slug, options: options)
            logHostUI("[HostUI] connected to \(daemon.binaryName) on \(connectionTarget())")
            isRunning = true
        } catch {
            daemonIdentity = nil
            cleanupStageRoot()
            connectionError = "Failed to start bundled daemon: \(String(describing: error))"
            isRunning = false
        }
#else
        connectionError = GreetingClientError.unsupportedPlatform.localizedDescription
        isRunning = false
#endif
    }

    func stop() {
        let currentClient = client
        client = nil
#if os(macOS)
        daemonIdentity = nil
#endif

        do {
            try currentClient?.close()
        } catch {
            connectionError = "Failed to stop daemon connection: \(error.localizedDescription)"
        }

#if os(macOS)
        cleanupStageRoot()
#endif
        isRunning = false
    }

    func listLanguages() async throws -> [Language] {
        if client == nil { start() }
        guard let client else {
            throw DaemonError.notConnected
        }
        return try await client.listLanguages()
    }

    func sayHello(name: String, langCode: String) async throws -> String {
        guard let client else {
            throw DaemonError.notConnected
        }
        return try await client.sayHello(name: name, langCode: langCode)
    }

    deinit {
        try? client?.close()
#if os(macOS)
        let root = stageRoot
        stageRoot = nil
        daemonIdentity = nil
        if let root {
            try? FileManager.default.removeItem(at: root)
        }
#endif
    }

    var assemblyFamily: String {
        let value = ProcessInfo.processInfo.environment["OP_ASSEMBLY_FAMILY"]?.trimmingCharacters(in: .whitespacesAndNewlines)
        return (value?.isEmpty == false ? value! : "Greeting-Swiftui-Go")
    }

    @Published var transport: String = {
        let value = ProcessInfo.processInfo.environment["OP_ASSEMBLY_TRANSPORT"]?.trimmingCharacters(in: .whitespacesAndNewlines)
        return (value?.isEmpty == false ? value! : "stdio")
    }() {
        didSet {
            // When transport changes, cleanly stop so the UI can trigger a full reload
            stop()
        }
    }

    var daemonBinaryName: String {
#if os(macOS)
        return daemonIdentity?.binaryName ?? (try? resolveDaemon().binaryName) ?? "gudule-daemon-greeting-swift"
#else
        return "gudule-daemon-greeting-swift"
#endif
    }

    private func connectionTarget() -> String {
        transport == "stdio" ? "stdio" : transport
    }

    private func logHostUI(_ line: String) {
        guard let data = (line + "\n").data(using: .utf8) else {
            return
        }
        FileHandle.standardError.write(data)
    }
}

enum DaemonError: LocalizedError {
    case notConnected

    var errorDescription: String? {
        switch self {
        case .notConnected:
            return "Not connected to the greeting daemon"
        }
    }
}

#if os(macOS)
private struct GreetingDaemonIdentity {
    static let binaryPrefix = "gudule-daemon-greeting-"

    let slug: String
    let familyName: String
    let binaryName: String
    let buildRunner: String
    let binaryPath: String

    static func fromBinaryPath(_ path: String) -> GreetingDaemonIdentity? {
        fromBinaryName(URL(fileURLWithPath: path).lastPathComponent, binaryPath: path)
    }

    static func fromBinaryName(_ binaryName: String, binaryPath: String) -> GreetingDaemonIdentity? {
        let normalized = binaryName.hasSuffix(".exe")
            ? String(binaryName.dropLast(4))
            : binaryName
        guard normalized.hasPrefix(binaryPrefix) else {
            return nil
        }

        let variant = String(normalized.dropFirst(binaryPrefix.count))
        return GreetingDaemonIdentity(
            slug: "gudule-greeting-daemon-\(variant)",
            familyName: "Greeting-Daemon-\(displayVariant(variant))",
            binaryName: normalized,
            buildRunner: runner(for: variant),
            binaryPath: URL(fileURLWithPath: binaryPath).standardizedFileURL.path
        )
    }

    private static func displayVariant(_ variant: String) -> String {
        let overrides: [String: String] = [
            "cpp": "CPP",
            "js": "JS",
            "qt": "Qt",
        ]

        return variant
            .split(separator: "-")
            .map { token in
                let value = String(token)
                if let override = overrides[value] {
                    return override
                }
                guard let first = value.first else {
                    return value
                }
                return String(first).uppercased() + value.dropFirst()
            }
            .joined(separator: "-")
    }

    private static func runner(for variant: String) -> String {
        switch variant {
        case "go":
            return "go-module"
        case "rust":
            return "cargo"
        case "swift":
            return "swift-package"
        case "kotlin":
            return "gradle"
        case "dart":
            return "dart"
        case "python":
            return "python"
        case "csharp":
            return "dotnet"
        case "node":
            return "npm"
        default:
            return "go-module"
        }
    }
}

private extension DaemonProcess {
    static let holonUUID = "61aa59e8-e4fc-425f-a799-48ff7a6d02d2"

    func resolveDaemon() throws -> GreetingDaemonIdentity {
        for candidate in daemonCandidates() {
            if let daemon = GreetingDaemonIdentity.fromBinaryPath(candidate),
               FileManager.default.isExecutableFile(atPath: daemon.binaryPath) {
                return daemon
            }
        }
        throw DaemonStartError.binaryNotFound(GreetingDaemonIdentity.binaryPrefix + "*")
    }

    func daemonCandidates() -> [String] {
        let fileManager = FileManager.default
        let currentDirectory = URL(
            fileURLWithPath: fileManager.currentDirectoryPath,
            isDirectory: true
        )

        var directories: [URL] = [
            currentDirectory.appendingPathComponent("build", isDirectory: true),
            currentDirectory.appendingPathComponent("../build", isDirectory: true),
            currentDirectory.appendingPathComponent("../../daemons", isDirectory: true),
        ]

        if let executableURL = Bundle.main.executableURL {
            let executableDir = executableURL.deletingLastPathComponent()
            directories.append(executableDir)
            directories.append(executableDir.appendingPathComponent("daemon", isDirectory: true))
            directories.append(executableDir.appendingPathComponent("../Resources", isDirectory: true))
            directories.append(executableDir.appendingPathComponent("../Resources/daemon", isDirectory: true))
        }

        if let resourceURL = Bundle.main.resourceURL {
            directories.append(resourceURL)
            directories.append(resourceURL.appendingPathComponent("daemon", isDirectory: true))
        }

        var seen = Set<String>()
        var candidates: [String] = []
        for directory in directories {
            let normalizedDir = directory.standardizedFileURL.path
            guard seen.insert(normalizedDir).inserted else {
                continue
            }

            var isDirectory: ObjCBool = false
            guard fileManager.fileExists(atPath: normalizedDir, isDirectory: &isDirectory),
                  isDirectory.boolValue else {
                continue
            }

            if directory.lastPathComponent == "daemons" {
                appendSourceTreeDaemonCandidates(from: directory, into: &candidates)
                continue
            }

            appendBundledBinaries(from: directory, into: &candidates)
        }

        return dedupeSortedPaths(candidates)
    }

    func appendBundledBinaries(from directory: URL, into results: inout [String]) {
        guard let entries = try? FileManager.default.contentsOfDirectory(
            at: directory,
            includingPropertiesForKeys: [.isRegularFileKey],
            options: [.skipsHiddenFiles]
        ) else {
            return
        }

        for entry in entries where entry.lastPathComponent.hasPrefix(GreetingDaemonIdentity.binaryPrefix) {
            results.append(entry.standardizedFileURL.path)
        }
    }

    func appendSourceTreeDaemonCandidates(from daemonsDir: URL, into results: inout [String]) {
        guard let entries = try? FileManager.default.contentsOfDirectory(
            at: daemonsDir,
            includingPropertiesForKeys: [.isDirectoryKey],
            options: [.skipsHiddenFiles]
        ) else {
            return
        }

        for entry in entries where entry.lastPathComponent.hasPrefix(GreetingDaemonIdentity.binaryPrefix) {
            results.append(
                entry
                    .appendingPathComponent(".op/build/bin", isDirectory: true)
                    .appendingPathComponent(entry.lastPathComponent)
                    .standardizedFileURL
                    .path
            )
            results.append(entry.appendingPathComponent(entry.lastPathComponent).standardizedFileURL.path)
        }
    }

    func dedupeSortedPaths(_ paths: [String]) -> [String] {
        var seen = Set<String>()
        return paths
            .map { URL(fileURLWithPath: $0).standardizedFileURL.path }
            .filter { seen.insert($0).inserted }
            .sorted()
    }

    func stageHolonRoot(daemon: GreetingDaemonIdentity) throws -> URL {
        let root = FileManager.default.temporaryDirectory
            .appendingPathComponent("gudule-greeting-hostui-swiftui-\(UUID().uuidString)", isDirectory: true)

        do {
            let holonDir = root
                .appendingPathComponent("holons", isDirectory: true)
                .appendingPathComponent(daemon.slug, isDirectory: true)
            try FileManager.default.createDirectory(at: holonDir, withIntermediateDirectories: true)
            try manifest(for: daemon)
                .write(to: holonDir.appendingPathComponent("holon.yaml"), atomically: true, encoding: .utf8)
            return root
        } catch {
            try? FileManager.default.removeItem(at: root)
            throw DaemonStartError.failedToStageRoot(error.localizedDescription)
        }
    }

    func manifest(for daemon: GreetingDaemonIdentity) -> String {
        """
        schema: holon/v0
        uuid: "\(Self.holonUUID)"
        given_name: "gudule"
        family_name: "\(daemon.familyName)"
        motto: "Greets users in 56 languages."
        composer: "Codex"
        clade: deterministic/pure
        status: draft
        born: "2026-03-12"
        generated_by: manual
        kind: native
        build:
          runner: \(daemon.buildRunner)
        artifacts:
          binary: "\(yamlEscape(daemon.binaryPath))"
        """ + "\n"
    }

    func yamlEscape(_ value: String) -> String {
        value.replacingOccurrences(of: "\\", with: "\\\\")
            .replacingOccurrences(of: "\"", with: "\\\"")
    }

    func cleanupStageRoot() {
        guard let root = stageRoot else { return }
        stageRoot = nil
        try? FileManager.default.removeItem(at: root)
    }
}

private enum DaemonStartError: LocalizedError {
    case binaryNotFound(String)
    case failedToStageRoot(String)
    case failedToEnterRoot(String)

    var errorDescription: String? {
        switch self {
        case let .binaryNotFound(binaryName):
            return "Daemon binary not found: \(binaryName)"
        case let .failedToStageRoot(message):
            return "Failed to stage holon root: \(message)"
        case let .failedToEnterRoot(path):
            return "Failed to enter staged holon root: \(path)"
        }
    }
}
#endif
